### PR TITLE
added plugin support

### PR DESCRIPTION
**Pull request by pjordaan**
Right now the functionality is hardcoded in Packer.php. With this the configuration is a little bit more general and new plugins can be added

### CODE REVIEW

File: src/ConfigInterface.php [hunk 1]
```diff
@@ -47,6 +49,13 @@ public function getEntryPoints(): array;
     public function getAssetFiles(): array;
 
     /**
+     * Returns a list of plugins.
+     *
+     * @return PluginInterface
```
Comment (nicoschoenmaker) [inline src/ConfigInterface.php]: PluginInterface[]

File: src/Packer.php [hunk 2]
```diff
@@ -75,42 +69,22 @@ public static function pack(string $project_root, LoggerInterface $logger, bool
...
 
-        // LESS
-        if ($config->isLessEnabled()) {
-            $less_collector = new LessImportCollector();
-            if ($config->isDev()) {
-                $less_collector = new CachedImportCollector($less_collector, $cache);
+        foreach ($config->getPlugins() as $plugin_class_name) {
+            $plugin = new $plugin_class_name($config, $cache, $nodejs);
+            if (!$plugin instanceof PluginInterface) {
+                throw new \RuntimeException("BLAH");
```
Comment (nicoschoenmaker) [inline src/Packer.php]: BLAH?

File: src/Plugin/LessPlugin.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+use Hostnet\Component\Resolver\Bundler\Processor\LessContentProcessor;
+use Hostnet\Component\Resolver\Bundler\Runner\LessRunner;
+use Hostnet\Component\Resolver\Cache\CachedImportCollector;
+use Hostnet\Component\Resolver\ConfigInterface;
+use Hostnet\Component\Resolver\Import\BuiltIn\LessImportCollector;
+use Hostnet\Component\Resolver\Import\Nodejs\Executable;
+use Psr\SimpleCache\CacheInterface;
+
+class LessPlugin implements PluginInterface
```
Comment (nicoschoenmaker) [inline src/Plugin/LessPlugin.php]: If it doesn't expose any methods that are not in the interface, it should be final (so.. all plugins?)

Comment (nicoschoenmaker) [inline src/Plugin/LessPlugin.php]: There should be some docs explaining which less we mean, what problem this solves, which node package that is, maybe a link to the website of less, ...

Ditto for other plugins

File: src/Plugin/EventSubscriberInterface.php [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+ */
+declare(strict_types=1);
+
+namespace Hostnet\Component\Resolver\Plugin;
+
+use Hostnet\Component\Resolver\Bundler\Pipeline\ContentPipelineInterface;
+use Hostnet\Component\Resolver\Import\ImportFinderInterface;
+
+/**
+ * Plugins should implement this interface if they require events attached to the pipeline.
```
Comment (nicoschoenmaker) [inline src/Plugin/EventSubscriberInterface.php]: We should probably denote an example of what you can do with these events here

It also would help devs a lot to link to the related Symfony docs!

File: test/FileConfigTest.php [hunk 5]
```diff
@@ -67,8 +63,6 @@ public function testMinimalRelativeNodePaths()
         self::assertSame(__DIR__ . '/fixtures/configs/var', $config->getCacheDir());
         self::assertSame(__DIR__ . '/fixtures/configs/bin/node', $config->getNodeJsBinary());
         self::assertSame(__DIR__ . '/fixtures/configs/node_modules', $config->getNodeModulesPath());
-        self::assertFalse($config->isLessEnabled());
-        self::assertFalse($config->isTsEnabled());
-        self::assertFalse($config->isAngularEnabled());
+        self::assertSame([], $config->getPlugins());
     }
 }
```
Comment (nicoschoenmaker) [inline test/FileConfigTest.php]: Missing a test-case where a plugin is activated.

File: src/Config/FileConfig.php [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    /**
+     * Location of the 'node_modules' folder.
+     *
+     * @return string
+     */
+    private function getNodeModulesPath(): string
+    {
+        if (File::isAbsolutePath($this->config_file_contents['node']['node_modules'])) {
```
Comment (nicoschoenmaker) [inline src/Config/FileConfig.php]:
```php
return File::makeAbsolutePath($this->config_file_contents['node']['node_modules'], $this->getProjectRoot());
```

ditto for getNodeJsBinary

File: src/Bundler/Pipeline/ContentPipeline.php [hunk 7]
```diff
@@ -26,7 +26,7 @@
  * pushed, it will go through various processors until it is written to disk as
  * the given output file.
  */
-final class ContentPipeline implements ContentPipelineInterface
+class ContentPipeline implements ContentPipelineInterface
```
Comment (iltar) [inline src/Bundler/Pipeline/ContentPipeline.php]: final should remain here, shouldn't create an extension point for 3rd parties as they can extend it, which makes maintenance a lot harder if we would ever change something.

Solution would be to create a second interface and use this for the adders.

File: src/Config/FileConfig.php [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+     */
+    public function __construct(
+        string $config_file,
+        array $plugins = [],
+        bool $dev = false,
+        EventDispatcherInterface $dispatcher = null,
+        LoggerInterface $logger = null
+    ) {
+        $this->config_file          = $config_file;
+        $this->config_file_contents = json_decode(file_get_contents($this->config_file), true);
```
Comment (iltar) [inline src/Config/FileConfig.php]: Not sure how safe this is, what if there's invalid json or the file does not exist?

File: src/Import/ImportFinder.php [hunk 9]
```diff
@@ -11,7 +11,7 @@
 /**
  * Import finder which uses ImportCollectorInterface to find imports.
  */
-final class ImportFinder implements ImportFinderInterface
+class ImportFinder implements ImportFinderInterface
```
Comment (iltar) [inline src/Import/ImportFinder.php]: Same as the other one, should probably not become an extension point. People _will_ extend and break it.

File: src/Plugin/AngularPlugin.php [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        $angular_collector = new AngularImportCollector();
+        if ($plugin_api->getConfig()->isDev()) {
+            $angular_collector = new CachedImportCollector($angular_collector, $plugin_api->getCache());
+        }
+        $plugin_api->addCollector($angular_collector);
+        $listener = new AngularHtmlListener($plugin_api);
+        $plugin_api->getConfig()->getEventDispatcher()->addListener(
+            AssetEvents::POST_PROCESS,
+            [$listener, 'onPostTranspile']
```
Comment (iltar) [inline src/Plugin/AngularPlugin.php]: variable could in theory be inlined as it's only used once

File: src/Plugin/PluginInterface.php [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+ * if ($plugin_api->getConfig()->isDev()) {
+ *     $angular_collector = new CachedImportCollector($angular_collector, $plugin_api->getCache());
+ * }
+ * $plugin_api->addCollector($angular_collector);
+ * ```
+ *
+ * Example: adding an event on post processing generated assets.
+ * ```
+ * $listener = new UglifyjsListener();
+ * $plugin_api->getConfig()->getEventDispatcher()->addListener(AssetEvents::POST_PROCESS, [$listener, 'onPostProcess']);
```
Comment (iltar) [inline src/Plugin/PluginInterface.php]: Small formatting thing, perhaps the notation should be changed slightly to allow it on multiple lines like in the actual code in our plugins

**Pull request merged**